controllers/appgroup_controller.go

```diff
@@ -499,32 +508,89 @@ func (r *ApplicationGroupReconciler) cleanupWorkflow(ctx context.Context, logr l
 	_ = r.List(ctx, &wfs, listOption)
 
 	if wfs.Items.Len() != 0 {
-		logr.Info("Reversing the workflow")
-
 		wf := wfs.Items[0]
 		for _, node := range wf.Status.Nodes {
 			nodes[node.ID] = node
 		}
-		graph, err := workflow.Build(g.Name, nodes)
+		rwf := &v1alpha12.Workflow{}
+
+		rwfName := fmt.Sprintf("%s-reverse", wf.Name)
+		rwfNamespace := wf.Namespace
+		err := r.Client.Get(ctx, types.NamespacedName{Namespace: rwfNamespace, Name: rwfName}, rwf)
 		if err != nil {
-			logr.Error(err, "failed to build the wf status DAG")
-		}
+			if kerrors.IsNotFound(err) {
+				logr.Info("Reversing the workflow")
+
+				err = r.generateReverseWorkflow(ctx, logr, nodes, &wf)
+				if err != nil {
+					logr.Error(err, "failed to generate reverse workflow")
+					// if generation of reverse workflow failed, delete the forward workflow and return
+					err = r.Client.Delete(ctx, &wf)
+					if err != nil {
+						logr.Error(err, "failed to delete workflow CRO")
+						return
+					}
+					return
+				}
+
+				wfPatch := client.MergeFrom(wf.DeepCopy())
 
-		rev := graph.Reverse()
+				// update the forward workflow metadata with a finalizer
+				controllerutil.AddFinalizer(&wf, rwffinalizer)
 
-		for _, bucket := range rev {
-			for _, hr := range bucket {
-				err = r.Client.Delete(ctx, &hr)
+				err = r.Client.Patch(ctx, &wf, wfPatch)
 				if err != nil {
-					logr.Error(err, "failed to delete helmrelease CRO - continuing with cleanup")
+					logr.Error(err, "failed to patch workflow CRO")
+					return
 				}
+
+				err = r.Client.Delete(ctx, &wf)
+				if err != nil {
+					logr.Error(err, "failed to delete workflow CRO - continuing with cleanup")
+					return
+				}
+
+				// reverse workflow started - requeue
+				requeue = true
```

Comment:
Can we be explicit with our return statements, setting the variable and then returning like this almost hides the behavior a bit in my opinion